File: src/android/app/src/main/java/com/founditure/presentation/components/FounditureButton.kt
/*
 * Human Tasks:
 * 1. Test button touch targets with Android's accessibility scanner
 * 2. Verify color contrast ratios meet WCAG 2.1 AA standards (4.5:1)
 * 3. Test button states (enabled, disabled, loading) across different devices
 * 4. Validate button behavior with screen readers and TalkBack
 */

package com.founditure.presentation.components

// androidx.compose.material3:material3:1.1.0
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.CircularProgressIndicator

// androidx.compose.runtime:runtime:1.5.0
import androidx.compose.runtime.Composable

// androidx.compose.foundation:foundation:1.5.0
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.width

// androidx.compose.ui:ui:1.5.0
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.material3.Text

// Internal theme imports
import com.founditure.presentation.theme.FounditureTheme

/**
 * A customizable Material3 button component that maintains consistent styling across the app
 * with WCAG 2.1 AA compliant touch targets and color contrast.
 *
 * Requirements addressed:
 * - Component Library (3.1.1): Custom Design System with atomic design principles
 * - Accessibility (3.1.1): Touch targets 44x44pt with WCAG 2.1 AA compliance
 * - Visual Hierarchy (3.1.1): Material Design 3 with dynamic color system
 *
 * @param text The button text to display
 * @param onClick Callback to be invoked when the button is clicked
 * @param enabled Whether the button is enabled
 * @param loading Whether to show a loading indicator
 */
@Composable
fun FounditureButton(
    text: String,
    onClick: () -> Unit,
    enabled: Boolean = true,
    loading: Boolean = false
) {
    Button(
        onClick = onClick,
        enabled = enabled && !loading,
        modifier = Modifier
            .height(44.dp) // WCAG 2.1 AA compliant touch target
            .width(200.dp),
        colors = ButtonDefaults.buttonColors(
            containerColor = FounditureTheme.colors.primary,
            contentColor = FounditureTheme.colors.onPrimary,
            disabledContainerColor = FounditureTheme.colors.primary.copy(alpha = 0.38f),
            disabledContentColor = FounditureTheme.colors.onPrimary.copy(alpha = 0.38f)
        ),
        shape = FounditureTheme.shapes.small
    ) {
        Box(
            contentAlignment = Alignment.Center,
            modifier = Modifier.padding(horizontal = 16.dp)
        ) {
            if (loading) {
                CircularProgressIndicator(
                    modifier = Modifier.size(24.dp),
                    color = FounditureTheme.colors.onPrimary,
                    strokeWidth = 2.dp
                )
            } else {
                Text(
                    text = text,
                    style = FounditureTheme.typography.labelLarge,
                    textAlign = TextAlign.Center
                )
            }
        }
    }
}

/**
 * An outlined variant of the Founditure button component with WCAG 2.1 AA compliant styling.
 *
 * Requirements addressed:
 * - Component Library (3.1.1): Custom Design System with atomic design principles
 * - Accessibility (3.1.1): Touch targets 44x44pt with WCAG 2.1 AA compliance
 * - Visual Hierarchy (3.1.1): Material Design 3 with dynamic color system
 *
 * @param text The button text to display
 * @param onClick Callback to be invoked when the button is clicked
 * @param enabled Whether the button is enabled
 */
@Composable
fun FounditureOutlinedButton(
    text: String,
    onClick: () -> Unit,
    enabled: Boolean = true
) {
    OutlinedButton(
        onClick = onClick,
        enabled = enabled,
        modifier = Modifier
            .height(44.dp) // WCAG 2.1 AA compliant touch target
            .width(200.dp),
        colors = ButtonDefaults.outlinedButtonColors(
            contentColor = FounditureTheme.colors.primary,
            disabledContentColor = FounditureTheme.colors.primary.copy(alpha = 0.38f)
        ),
        shape = FounditureTheme.shapes.small
    ) {
        Text(
            text = text,
            style = FounditureTheme.typography.labelLarge,
            textAlign = TextAlign.Center,
            modifier = Modifier.padding(horizontal = 16.dp)
        )
    }
}

File: src/ios/Founditure/UI/Components/FounditureButton.swift
//
// FounditureButton.swift
// Founditure
//
// Human Tasks:
// 1. Verify button touch target sizes meet WCAG requirements on all supported devices
// 2. Test color contrast ratios in both light and dark modes
// 3. Validate dynamic type scaling across all text size categories
// 4. Review haptic feedback implementation with UX team
// 5. Test VoiceOver compatibility and accessibility labels

import SwiftUI // Latest
import Foundation

// Import theme dependencies
import "../Theme/Colors"
import "../Theme/Typography"

/// Button style options following Material Design 3 guidelines
/// Requirement: Visual Hierarchy - Material Design 3 with dynamic color system
public enum ButtonStyle {
    case primary
    case secondary
    case text
}

/// Button size options with WCAG-compliant touch targets
/// Requirement: Accessibility - WCAG 2.1 AA compliance with minimum touch targets 44x44pt
public enum ButtonSize {
    case small
    case regular
    case large
}

/// A customizable button component that follows the Founditure design system
/// Requirements addressed:
/// - Visual Hierarchy: Material Design 3 with 8dp grid system and dynamic color system
/// - Component Library: Custom Design System with atomic design principles
/// - Accessibility: WCAG 2.1 AA compliance with minimum touch targets 44x44pt
@available(iOS 14.0, *)
public struct FounditureButton: View {
    // MARK: - Properties
    
    private let title: String
    private let style: ButtonStyle
    private let size: ButtonSize
    private let isEnabled: Bool
    private let isLoading: Bool
    private let action: (() -> Void)?
    
    // MARK: - Private Properties
    
    private let minTouchTargetSize: CGFloat = 44 // WCAG 2.1 AA requirement
    
    // MARK: - Initialization
    
    /// Creates a new FounditureButton with the specified parameters
    /// - Parameters:
    ///   - title: The button's text label
    ///   - style: The button's visual style
    ///   - size: The button's size category
    ///   - isEnabled: Whether the button is interactive
    ///   - isLoading: Whether to show a loading indicator
    ///   - action: The closure to execute when tapped
    public init(
        title: String,
        style: ButtonStyle = .primary,
        size: ButtonSize = .regular,
        isEnabled: Bool = true,
        isLoading: Bool = false,
        action: (() -> Void)? = nil
    ) {
        self.title = title
        self.style = style
        self.size = size
        self.isEnabled = isEnabled
        self.isLoading = isLoading
        self.action = action
    }
    
    // MARK: - Body
    
    public var body: some View {
        Button(action: {
            if isEnabled && !isLoading {
                // Add haptic feedback for button press
                let generator = UIImpactFeedbackGenerator(style: .medium)
                generator.impactOccurred()
                
                action?()
            }
        }) {
            HStack(spacing: 8) {
                if isLoading {
                    ProgressView()
                        .progressViewStyle(CircularProgressViewStyle(
                            tint: style == .primary ? FounditureColors.onPrimary : FounditureColors.primary
                        ))
                }
                
                Text(title)
                    .font(FounditureTypography.dynamicFont(
                        style: .body,
                        size: .medium
                    ))
            }
            .frame(maxWidth: size == .small ? nil : .infinity)
            .frame(minWidth: minTouchTargetSize, minHeight: minTouchTargetSize)
            .padding(.horizontal, buttonHorizontalPadding)
            .padding(.vertical, buttonVerticalPadding)
            .background(buttonBackground)
            .foregroundColor(buttonForegroundColor)
            .cornerRadius(8)
            .opacity(isEnabled ? 1.0 : 0.5)
        }
        .buttonStyle(PlainButtonStyle())
        .disabled(!isEnabled || isLoading)
        .accessibilityLabel(title)
        .accessibilityHint(isEnabled ? nil : "Button is disabled")
        .accessibilityAddTraits(.isButton)
        .if(isLoading) { view in
            view.accessibilityValue("Loading")
        }
    }
    
    // MARK: - Private Helpers
    
    private var buttonHorizontalPadding: CGFloat {
        switch size {
        case .small:
            return 12
        case .regular:
            return 16
        case .large:
            return 24
        }
    }
    
    private var buttonVerticalPadding: CGFloat {
        switch size {
        case .small:
            return 8
        case .regular:
            return 12
        case .large:
            return 16
        }
    }
    
    private var buttonBackground: Color {
        switch style {
        case .primary:
            return FounditureColors.dynamicColor(
                lightColor: FounditureColors.primary,
                darkColor: FounditureColors.primary,
                colorScheme: .light
            )
        case .secondary:
            return FounditureColors.dynamicColor(
                lightColor: FounditureColors.secondary,
                darkColor: FounditureColors.secondary,
                colorScheme: .light
            )
        case .text:
            return .clear
        }
    }
    
    private var buttonForegroundColor: Color {
        switch style {
        case .primary, .secondary:
            return FounditureColors.onPrimary
        case .text:
            return FounditureColors.primary
        }
    }
}

// MARK: - View Modifier Extension

extension View {
    fileprivate func `if`<Content: View>(_ condition: Bool, transform: (Self) -> Content) -> AnyView {
        if condition {
            return AnyView(transform(self))
        }
        return AnyView(self)
    }
}

#if DEBUG
// MARK: - Preview Provider

struct FounditureButton_Previews: PreviewProvider {
    static var previews: some View {
        VStack(spacing: 20) {
            FounditureButton(
                title: "Primary Button",
                style: .primary,
                action: {}
            )
            
            FounditureButton(
                title: "Secondary Button",
                style: .secondary,
                action: {}
            )
            
            FounditureButton(
                title: "Text Button",
                style: .text,
                action: {}
            )
            
            FounditureButton(
                title: "Loading Button",
                isLoading: true,
                action: {}
            )
            
            FounditureButton(
                title: "Disabled Button",
                isEnabled: false,
                action: {}
            )
        }
        .padding()
        .previewLayout(.sizeThatFits)
    }
}
#endif